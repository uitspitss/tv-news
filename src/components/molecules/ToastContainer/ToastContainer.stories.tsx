import React, { VFC } from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import base from 'paths.macro';
import { toast } from 'react-toastify';

import { ToastContainer, ToastContainerProps } from './ToastContainer';

const path = base.replace(
  /^.+\/(atoms|molecules|organisms|templates|pages)\/(.+)\//,
  '$1/$2',
);

export default {
  title: `${path}`,
  component: ToastContainer,
} as Meta;

const Component: VFC<ToastContainerProps> = (args) => (
  <>
    <ToastContainer {...args} />
    <button type="button" onClick={() => toast.success('toast success')}>
      success toast
    </button>
    <button type="button" onClick={() => toast.info('toast info')}>
      info toast
    </button>
    <button type="button" onClick={() => toast.warn('toast warn')}>
      warn toast
    </button>
    <button type="button" onClick={() => toast.error('toast error')}>
      error toast
    </button>
  </>
);

const Template: Story<ToastContainerProps> = (args) => <Component {...args} />;

export const Default = Template.bind({});
