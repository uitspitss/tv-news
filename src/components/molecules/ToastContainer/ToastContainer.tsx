import React, { memo, VFC } from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import tw, { styled } from 'twin.macro';
import {
  ToastContainer as ReactToastifyContainer,
  Slide,
} from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';

export type ToastContainerProps = {
  /**
   * no className
   */
  className?: undefined;
};

type Props = ToastContainerProps;

const Component: VFC<Props> = (props) => (
  <ReactToastifyContainer
    position="top-right"
    autoClose={5000}
    closeOnClick
    closeButton
    draggable={false}
    hideProgressBar
    transition={Slide}
    {...props}
  />
);

const StyledComponent = styled(Component)`
  ${tw`fixed top-12`}

  .Toastify__toast {
    ${tw`flex items-center p-2 cursor-pointer mb-1`}
  }

  .Toastify__toast--success {
    ${tw`text-success-500 bg-success-100`}
  }

  .Toastify__toast--info {
    ${tw`text-info-500 bg-info-100`}
  }

  .Toastify__toast--warning {
    ${tw`text-warning-500 bg-warning-100`}
  }

  .Toastify__toast--error {
    ${tw`text-error-500 bg-error-100`}
  }
`;

/**
 * ToastContainer of react-toastify
 */
export const ToastContainer = memo((props: ToastContainerProps) => (
  <StyledComponent {...props} />
));
