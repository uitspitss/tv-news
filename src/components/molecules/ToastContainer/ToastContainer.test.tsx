import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { toast } from 'react-toastify';

import { ToastContainer } from './ToastContainer';

describe('molecules/ToastContainer', () => {
  test('snapshot test', () => {
    const { container } = render(
      <>
        <ToastContainer />
        <button type="button" onClick={() => toast.error('error')}>
          trigger
        </button>
      </>,
    );
    expect(container).toMatchSnapshot();
  });

  describe('Behavior', () => {
    beforeEach(() => {
      jest.clearAllMocks();
      render(
        <>
          <ToastContainer />
          <button type="button" onClick={() => toast.error('error')}>
            trigger error
          </button>
          <button type="button" onClick={() => toast.success('success')}>
            trigger success
          </button>
          <button type="button" onClick={() => toast.info('info')}>
            trigger info
          </button>
          <button type="button" onClick={() => toast.warn('warn')}>
            trigger warn
          </button>
        </>,
      );
    });

    test('initial view', () => {
      expect(
        screen.getByRole('button', { name: 'trigger error' }),
      ).toBeInTheDocument();
    });

    test('click, then error toast', async () => {
      fireEvent.click(screen.getByRole('button', { name: 'trigger error' }));

      expect(await screen.findByRole('alert')).toBeInTheDocument();
      expect(await screen.findByText('error')).toBeInTheDocument();
    });

    test('click, then success toast', async () => {
      fireEvent.click(screen.getByRole('button', { name: 'trigger success' }));

      expect(await screen.findByRole('alert')).toBeInTheDocument();
      expect(await screen.findByText('success')).toBeInTheDocument();
    });

    test('click, then info toast', async () => {
      fireEvent.click(screen.getByRole('button', { name: 'trigger info' }));

      expect(await screen.findByRole('alert')).toBeInTheDocument();
      expect(await screen.findByText('info')).toBeInTheDocument();
    });

    test('click, then warn toast', async () => {
      fireEvent.click(screen.getByRole('button', { name: 'trigger warn' }));

      expect(await screen.findByRole('alert')).toBeInTheDocument();
      expect(await screen.findByText('warn')).toBeInTheDocument();
    });
  });
});
