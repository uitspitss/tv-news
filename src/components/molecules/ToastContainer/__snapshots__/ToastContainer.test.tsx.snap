// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`molecules/ToastContainer snapshot test 1`] = `
<div>
  <div
    class="Toastify"
  />
  <button
    type="button"
  >
    trigger
  </button>
</div>
`;
